### ISSUE TITLE
[Feature Request] Refactor offline compilation download of files

### ISSUE THREAD

**Issue by akohlmey** [lammps/lammps#3282]
**Summary**

The current handling of transparent downloads is in several cases dependent on the compilation host being connected to the internet. There are some ways to provide already downloaded archives, but this has known problems, e.g. updates are not applied due to the variables being cached. Also there should be a test whether downloads are possible and CMake should not attempt a download in that case. If offline compilation is intended, this should be indicated and a folder path with previously downloaded source archives provided. The offline use path can then be inferred from the original download URL.

In summary:
- download URLs and hashes should not be stored in cached variables
- implement a test to detect if downloads are possible or not. the CMake scripting should react accordingly and error out or skip steps
- use a single variable to point to a folder with pre-downloaded sources instead of requiring to override all URLs. URLs will be adapted by CMake in offline mode

**Comment by ocaisa** [comment]
Just to add to this, the current approach for building the docs also requires access to the internet...there is no (offline) way to provide the necessary packages to the `venv` used.

**Comment by akohlmey** [comment]
> Just to add to this, the current approach for building the docs also requires access to the internet...there is no (offline) way to provide the necessary packages to the `venv` used.

This is not likely to change (that part of the build system is already complex as it is to be (more) portable). The documentation can be built anywhere and then only the resulting files (html folder, pdf or epub file) need to be transferred to the target location (which may be without internet access).

This is different from compiling executables for which it may only be possible to compile them on a disconnected machine.

**Comment by mcepl** [comment]
So, when the build for openSUSE in our network isolated environment fails with “Downloading external potential C_10_10.mesocnt from https://download.lammps.org/potentials” is it this bug or not?

**Comment by akohlmey** [comment]
> So, when the build for openSUSE in our network isolated environment fails with “Downloading external potential C_10_10.mesocnt from [https://download.lammps.org/potentials”](https://download.lammps.org/potentials%E2%80%9D) is it this bug or not?

You can prevent the download by using `-D DOWNLOAD_POTENTIALS=off` when configuring LAMMPS with CMake.
It appears that this setting is currently undocumented, so I will have to update the documentation. It currently affects only one potential file for a rarely used package, AFAICR.
